File: lint-arwaky/src-python/capabilities/dispatch_routing_parser.py
"""dispatch_routing_parser — Parser logic for capability class extraction."""

from __future__ import annotations
import re

from ..taxonomy import (
    ClassDefinitionMap,
    ClassMethodsVO,
    ContentString,
)
from .dispatch_parser_types import (
    BraceDepthVO,
    ClassParsingStateVO,
    ScopeNameVO,
)
from ..contract import IDispatchRoutingParserProtocol


class DispatchRoutingParser(IDispatchRoutingParserProtocol):
    """Helper to parse source code and extract class/method structures."""

    def strip_docstrings(self, text: ContentString) -> ContentString:
        """Remove triple-quoted strings and comments from text."""
        content = str(text.value)
        # Remove """...""" docstrings
        content = re.sub(r'""".*?"""', "", content, flags=re.DOTALL)
        # Remove '''...''' docstrings
        content = re.sub(r"'''[\s\S]*?'''", "", content)
        # Remove single-line comments
        content = re.sub(r"^\s*#.*$", "", content, flags=re.MULTILINE)
        return ContentString(value=content)

    def extract_class_methods(self, text: ContentString) -> ClassDefinitionMap:
        """Parse text to find class definitions and their methods."""
        result = ClassDefinitionMap()
        current_class = ScopeNameVO()
        class_brace_depth = BraceDepthVO()
        current_brace_depth = BraceDepthVO()

        for line in text.value.split("\n"):
            stripped = line.strip()
            state = self._process_class_line(
                ContentString(value=stripped),
                ContentString(value=line),
                ClassParsingStateVO(
                    current_class=current_class,
                    class_brace_depth=class_brace_depth,
                    current_brace_depth=current_brace_depth,
                ),
                result,
            )
            current_class = state.current_class
            class_brace_depth = state.class_brace_depth
            current_brace_depth = state.current_brace_depth

        return result

    def _process_class_line(
        self,
        stripped: ContentString,
        line: ContentString,
        state: ClassParsingStateVO,
        result: ClassDefinitionMap,
    ) -> ClassParsingStateVO:
        """Process a single line of text for class/method extraction."""
        current_class = state.current_class
        class_brace_depth = state.class_brace_depth
        current_brace_depth = state.current_brace_depth

        # Detect class definition
        class_result = self._handle_class_definition(
            stripped, current_class, result, class_brace_depth, current_brace_depth
        )
        if class_result is not None:
            return class_result

        # Detect method definition inside a class
        if current_class.value is not None:
            self._handle_method_definition(stripped, line, current_class, result)

            # Track braces for scope tracking
            current_brace_depth.value += stripped.value.count(
                "{"
            ) - stripped.value.count("}")

            # Check if we left the class scope (dedented back to class level)
            current_class = self._handle_scope_exit(
                line, current_class, class_brace_depth, current_brace_depth
            )

        return ClassParsingStateVO(
            current_class=current_class,
            class_brace_depth=class_brace_depth,
            current_brace_depth=current_brace_depth,
        )

    def _handle_class_definition(
        self,
        stripped: ContentString,
        current_class: ScopeNameVO,
        result: ClassDefinitionMap,
        class_brace_depth: BraceDepthVO,
        current_brace_depth: BraceDepthVO,
    ) -> ClassParsingStateVO | None:
        """Detect a class definition line and update state."""
        class_match = re.match(r"^class\s+([A-Za-z_][\w]*)\s*[\(:]", stripped.value)
        if not class_match:
            return None
        new_class = ScopeNameVO(value=class_match.group(1))
        result.definitions[new_class.value] = ClassMethodsVO()
        class_brace_depth.value = current_brace_depth.value
        current_brace_depth.value += stripped.value.count("{") - stripped.value.count(
            "}"
        )
        return ClassParsingStateVO(
            current_class=new_class,
            class_brace_depth=class_brace_depth,
            current_brace_depth=current_brace_depth,
        )

    def _handle_method_definition(
        self,
        stripped: ContentString,
        line: ContentString,
        current_class: ScopeNameVO,
        result: ClassDefinitionMap,
    ) -> None:
        """Detect and record a method definition if found inside current class."""
        method_match = re.match(
            r"^(?:async\s+)?def\s+([A-Za-z_][\w]*)\s*\(", stripped.value
        )
        if method_match and current_class.value is not None:
            indent = len(line.value) - len(line.value.lstrip())
            if indent <= 8:
                result.definitions[current_class.value].methods.append(
                    method_match.group(1)
                )

    def _handle_scope_exit(
        self,
        line: ContentString,
        current_class: ScopeNameVO,
        class_brace_depth: BraceDepthVO,
        current_brace_depth: BraceDepthVO,
    ) -> ScopeNameVO:
        """Detect when we've left the class scope due to dedent."""
        stripped = line.value.strip()
        if stripped == "" or stripped.startswith("#"):
            return current_class

        # Check indentation of original line
        is_indented = line.value.startswith(" ") or line.value.startswith("\t")

        if not is_indented:
            if current_brace_depth.value <= class_brace_depth.value:
                if (
                    not stripped.startswith("@")
                    and not stripped.startswith("class ")
                    and not stripped.startswith("def ")
                    and not stripped.startswith("async ")
                ):
                    return ScopeNameVO(value=None)
        return current_class

File: lint-arwaky/src-rust/capabilities/dispatch_routing_parser.rs
// dispatch_routing_parser — Parser logic for capability class extraction.
// 1:1 Rust implementation matching capabilities/dispatch_routing_parser.py

use regex::Regex;
use once_cell::sync::Lazy;

use crate::taxonomy::{ClassDefinitionMap, ClassMethodsVO, ContentString};
use super::dispatch_parser_types::{BraceDepthVO, ClassParsingStateVO, ScopeNameVO};

pub struct DispatchRoutingParser;

impl DispatchRoutingParser {
    pub fn new() -> Self {
        Self
    }

    pub fn strip_docstrings(&self, text: &str) -> String {
        static DOUBLE_DOC_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r#""""[\s\S]*?""""#).unwrap()
        });
        static SINGLE_DOC_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r"'''[\s\S]*?'''").unwrap()
        });
        static COMMENT_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r"(?m)^\s*#.*$").unwrap()
        });

        let mut result = DOUBLE_DOC_RE.replace_all(text, "").to_string();
        result = SINGLE_DOC_RE.replace_all(&result, "").to_string();
        result = COMMENT_RE.replace_all(&result, "").to_string();
        result
    }

    pub fn extract_class_methods(&self, text: &str) -> ClassDefinitionMap {
        let mut result = ClassDefinitionMap {
            definitions: std::collections::HashMap::new(),
        };
        let mut current_class = ScopeNameVO::default();
        let mut class_brace_depth = BraceDepthVO::default();
        let mut current_brace_depth = BraceDepthVO::default();

        for line in text.lines() {
            let stripped = line.trim();
            let state = self._process_class_line(
                stripped,
                line,
                ClassParsingStateVO {
                    current_class: current_class.clone(),
                    class_brace_depth: class_brace_depth.clone(),
                    current_brace_depth: current_brace_depth.clone(),
                },
                &mut result,
            );
            current_class = state.current_class;
            class_brace_depth = state.class_brace_depth;
            current_brace_depth = state.current_brace_depth;
        }

        result
    }

    fn _process_class_line(
        &self,
        stripped: &str,
        line: &str,
        mut state: ClassParsingStateVO,
        result: &mut ClassDefinitionMap,
    ) -> ClassParsingStateVO {
        let class_result = self._handle_class_definition(
            stripped,
            &state.current_class,
            result,
            &state.class_brace_depth,
            &state.current_brace_depth,
        );
        if let Some(new_state) = class_result {
            return new_state;
        }

        if state.current_class.value.is_some() {
            self._handle_method_definition(stripped, line, &state.current_class, result);

            let open = stripped.chars().filter(|&c| c == '{').count();
            let close = stripped.chars().filter(|&c| c == '}').count();
            state.current_brace_depth.value += open as i32 - close as i32;

            state.current_class = self._handle_scope_exit(
                line,
                &state.current_class,
                &state.class_brace_depth,
                &state.current_brace_depth,
            );
        }

        ClassParsingStateVO {
            current_class: state.current_class,
            class_brace_depth: state.class_brace_depth,
            current_brace_depth: state.current_brace_depth,
        }
    }

    fn _handle_class_definition(
        &self,
        stripped: &str,
        current_class: &ScopeNameVO,
        result: &mut ClassDefinitionMap,
        _class_brace_depth: &BraceDepthVO,
        current_brace_depth: &BraceDepthVO,
    ) -> Option<ClassParsingStateVO> {
        static CLASS_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r"^class\s+([A-Za-z_][\w]*)\s*[\(:]").unwrap()
        });

        let caps = CLASS_RE.captures(stripped)?;
        let new_class_name = caps.get(1).unwrap().as_str().to_string();
        result.definitions.insert(new_class_name.clone(), ClassMethodsVO { methods: vec![] });

        let mut new_brace = current_brace_depth.clone();
        new_brace.value += stripped.chars().filter(|&c| c == '{').count() as i32
            - stripped.chars().filter(|&c| c == '}').count() as i32;

        Some(ClassParsingStateVO {
            current_class: ScopeNameVO { value: Some(new_class_name) },
            class_brace_depth: BraceDepthVO { value: current_brace_depth.value },
            current_brace_depth: new_brace,
        })
    }

    fn _handle_method_definition(
        &self,
        stripped: &str,
        line: &str,
        current_class: &ScopeNameVO,
        result: &mut ClassDefinitionMap,
    ) {
        static METHOD_RE: Lazy<Regex> = Lazy::new(|| {
            Regex::new(r"^(?:async\s+)?def\s+([A-Za-z_][\w]*)\s*\(").unwrap()
        });

        if let Some(caps) = METHOD_RE.captures(stripped) {
            if let Some(ref class_name) = current_class.value {
                let indent = line.len() - line.trim_start().len();
                if indent <= 8 {
                    if let Some(methods_vo) = result.definitions.get_mut(class_name) {
                        methods_vo.methods.push(caps.get(1).unwrap().as_str().to_string());
                    }
                }
            }
        }
    }

    fn _handle_scope_exit(
        &self,
        line: &str,
        current_class: &ScopeNameVO,
        class_brace_depth: &BraceDepthVO,
        current_brace_depth: &BraceDepthVO,
    ) -> ScopeNameVO {
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') {
            return current_class.clone();
        }

        let is_indented = line.starts_with(' ') || line.starts_with('\t');

        if !is_indented && current_brace_depth.value <= class_brace_depth.value {
            if !trimmed.starts_with('@')
                && !trimmed.starts_with("class ")
                && !trimmed.starts_with("def ")
                && !trimmed.starts_with("async ")
            {
                return ScopeNameVO { value: None };
            }
        }
        current_class.clone()
    }
}
